### DISCUSSION TITLE
Best practice for asynchronous state management?

### DISCUSSION THREAD

**Discussion by kiasta** [quasarframework/quasar#13456] [General - Components / Directives / etc] [answered]
I'm bad with titles, so let me try to explain. What I'm trying to do is create a state called "Categories" that the initial data is gathered by an API call to a database, like so:

[stores/categories.js](https://codepen.io/kiasta/pen/WNMpVpQ?editors=0010)
[MainLayout.vue](https://codepen.io/kiasta/pen/XWZRrrB?editors=0010)

![localvariableset](https://user-images.githubusercontent.com/8505162/169066003-41008358-fc65-411d-88a2-50d9f89b8c06.png)

As you can see from the image above, this "works" per say. The problem is that if I want to use the getters the UI does not update, so I have to set a local variable instead like what I'm doing in the above code. For instance, if I made this call without returning the categories state, or at least not consuming the returned value, like so:

[MainLayout.vue](https://codepen.io/kiasta/pen/oNEZKwP?editors=0010)
[CategoriesPage.vue](https://codepen.io/kiasta/pen/gOvWbwy?editors=0010)

![usingcategoriesstategetters jpg](https://user-images.githubusercontent.com/8505162/169066182-7ee3376b-a561-4222-82f5-d0f15dfd720d.png)
![categoriestable jpg](https://user-images.githubusercontent.com/8505162/169073183-60c5a187-612a-466e-84d0-f2d72e66df11.png)

I should expect that getAllCategories/getParentCategories/getNonParentCategories getters should return a populated list, regardless if the API call was async, right? What happens is I get an empty list instead and nothing rendered when the page loads. I can console.log inside of the getCategoriesFromServer function and the categories state is populated. And when I route to the categories page, the table that shows the categories is populated as expected. But only when I navigate to the page, not when I'm already there and reload, the table then becomes empty. The UI doesn't update or render the categories while it is on the pages that consume the state. It only renders after I navigate to said page. The menu never updates, no matter which page I route to.

So what I'm thinking is that the API call to the database is still waiting when the menu is rendered and there is no data, I'm guessing there is a way to wait for this data before rendering everything, but I just do not know how to do this. Or maybe I'm thinking about it the wrong way? State Management is a new concept for me and I've never used VueJS or Pinia before this project. Maybe someone can point me in the right direction? Much appreciated!

I'm using CodePen because the code formatter on here is so bad, I hope that is OK?

EDIT: Forgot to include the CategoriesPage.vue file and screenshot for context.

**Comment by MilosPaunovic** [comment]
It would be very useful to provide runnable code reproduction, either Codepen, Codesandbox or a Github repository, as this way it's hard for someone to jump into the problem and help you out.

**Reply by kiasta** [reply]
Yeah I get that, I assumed there was something I was doing wrong with the state and was overlooking something simple. I haven't played too much with codepen, but I'll see if I can make a working example when I get the chance. I appreciate the reply!

**Reply by kiasta** [reply]
@MilosPaunovic Hello, sorry for the late reply. Yesterday was quite busy for me and wasn't able to get around to uploading a test case. I added you to my private repo. I created a mock api as well and was able to reproduce the issue on my end.

To reproduce just load the page while in the "Dashboard" page. Then navigate to Admin->Categories. The table should populate, but the categories in the menu itself is not. Then reload the page while inside of the Categories page and you should see that the table is no longer populated. But the logs show data incoming.

To re-iterate, I'm not entirely sure what the best practice would be for doing something like this as I have no idea how to consume the async api call properly in this case. Hope this is clear, if not sorry.

**Reply by metalsadman** [reply]
Use computed property for your getters in your component.

**Reply by MilosPaunovic** [reply]
Categories table is loaded correctly, as it seems...
![Screenshot 2022-05-21 at 12 58 22](https://user-images.githubusercontent.com/62475782/169648598-2f03afc0-bddd-4104-9c10-c53b5e4043d5.png)

**Reply by kiasta** [reply]
Sorry to the both of you for the late reply, I had a procedure Monday and took a few days off work.

@MilosPaunovic Sorry, I was not clear. The categories in the menu on the left are supposed to be populated as well with the same categories as in the table. Also the table does not populate when you reload the browser while on that page. This is what happens when you reload the browser while on the categories page:

![Screenshot 2022-05-25 080335](https://user-images.githubusercontent.com/8505162/170257706-dd3a2278-b69b-4926-b1c7-5bbf43dd1081.png)


@metalsadman That doesn't seem to work either. I had computed variables, but it's just not being updating at all. I just retried using the computed variables and even renamed them and I'm still getting an empty page and when I throw a console.log in there to see the results of the getAllCategories getter, it's just an empty array. As you can see above, it's not updating even while using the computed variables.


<q-item
v-for="category in nonParentCategoriesComputed"
:key="category.id"
clickable
v-ripple
@click="$router.push({ name: 'Product', params: { category: category?.name?.toLowerCase() } })"
>


<q-expansion-item
v-for="category in parentCategoriesComputed"
:key="category.id"
:label="category.name"
icon="dashboard"
default-opened
>


computed: {
parentCategoriesComputed: function () {
return this.getAllCategories
},
nonParentCategoriesComputed: function () {
return this.getAllCategories
}
}

**Comment by MilosPaunovic** [comment]
For the drawer, inspect what you are looping through (`v-for="category in nonParentCategoriesComputed"`) and check what it returns. For the reload of page, you have to call your methods in mounted or some other lifecycle hook is you want to populate data.

Anyway, this is not a preffered way to get support from the community. You should isolate the problems and post 'em separatly included with a public reproduction of some kind, either Codepen, Codesanbox, Stackblitz, or GitHub repo, if there is no other way.

**Reply by kiasta** [reply]
You're right, I was having a hard time explaining what my issue was but I was just using pinia incorrectly. I got it all figured out. I tried to give you a shortened version as just an example, but it ended up being a lot of explaining and code snippets. Sorry about that. Thanks for the help actually, what you said made a lot of sense.

**Comment by TobyMosque** [comment, marked as answer]
u can't extract the state and getters of the store and expose them directly, u'll need wrap them first or expose the store directly,

that is good:
```html
<div>
  <div v-for="cat in store.categories" :key="cat.id">cat.name</div>
  <div v-for="cat in store.getParentCategories" :key="cat.id">cat.name</div>
  <q-btn click="store.addToCategories('new')"></q-btn>
</div>
```
```js
import { useCategories } from 'store/categories'

export default {
  setup () {
    const store = useCategories()
    return {
      store
    }
  }
}
```

but that will not work:
```html
<div>
  <div v-for="cat in categories" :key="cat.id">cat.name</div>
  <div v-for="cat in getParentCategories" :key="cat.id">cat.name</div>
  <q-btn click="addToCategories('new')"></q-btn>
</div>
```
```js
import { useCategories } from 'store/categories'

export default {
  setup () {
    const { categories, getParentCategories, addToCategories } = useCategories()
    return {
      categories,
      getParentCategories,
      addToCategories
    }
  }
}
```

how to fix that:
```html
<div>
  <div v-for="cat in categories" :key="cat.id">cat.name</div>
  <div v-for="cat in getParentCategories" :key="cat.id">cat.name</div>
  <q-btn click="addToCategories('new')"></q-btn>
</div>
```
```js
import { useCategories } from 'store/categories'
import { storeToRefs } from 'pinia'
import { computed } from 'vue'

export default {
  setup () {
    const store = useCategories()
    const { categories } = storeToRefs(store)
    const { addToCategories } = store
    const getParentCategories = computed(() => store.getParentCategories)
    return {
      categories,
      getParentCategories,
      addToCategories
    }
  }
}
```

**Reply by kiasta** [reply]
I didn't realize I was doing it wrong, I was just going by the examples of the documentation. But I got it all figured out thanks for your help!
